Riot/Managers/PushNotification/PushNotificationService.m

```diff
@@ -124,11 +151,104 @@ - (void)didReceiveRemoteNotification:(NSDictionary *)userInfo
 - (void)deregisterRemoteNotifications
 {
     _isPushRegistered = NO;
+    self.shouldReceiveVoIPPushes = NO;
+}
+
+- (void)applicationWillResignActive
+{
+    [[UNUserNotificationCenter currentNotificationCenter] removeUnwantedNotifications];
+    [[UNUserNotificationCenter currentNotificationCenter] removeCallNotificationsFor:nil];
+}
+
+- (void)applicationDidEnterBackground
+{
+    if (_pushNotificationManager.pushKitToken)
+    {
+        self.shouldReceiveVoIPPushes = YES;
+    }
 }
 
-- (void)applicationWillEnterForeground
+- (void)applicationDidBecomeActive
 {
     [[UNUserNotificationCenter currentNotificationCenter] removeUnwantedNotifications];
+    [[UNUserNotificationCenter currentNotificationCenter] removeCallNotificationsFor:nil];
+    if (_pushNotificationManager.pushKitToken)
+    {
+        self.shouldReceiveVoIPPushes = NO;
+    }
+}
+
+#pragma mark - Private Methods
+
+- (void)setShouldReceiveVoIPPushes:(BOOL)shouldReceiveVoIPPushes
+{
+    _shouldReceiveVoIPPushes = shouldReceiveVoIPPushes;
+    
+    if (_shouldReceiveVoIPPushes && _pushNotificationManager.pushKitToken)
+    {
+        MXSession *session = [AppDelegate theDelegate].mxSessions.firstObject;
+        if (session.state >= MXSessionStateStoreDataReady)
+        {
+            [self configurePushKit];
+        }
+        else
+        {
+            //  add an observer for session state
+            matrixSessionStateObserver = [[NSNotificationCenter defaultCenter] addObserverForName:kMXSessionStateDidChangeNotification object:nil queue:[NSOperationQueue mainQueue] usingBlock:^(NSNotification *notif) {
```

Comment:
We should do that in other classes of the project also but as pointed in Apple [documentation](https://developer.apple.com/documentation/foundation/nsnotificationcenter/1411723-addobserverforname) it's better from memory management pov if now we use:
```
@property (nonatomic, weak) id matrixSessionStateObserver;
```

then:

```
		   MXWeakify(self);

		   NSNotificationCenter * __weak notificationCenter = [NSNotificationCenter defaultCenter];
           matrixSessionStateObserver = [[NSNotificationCenter defaultCenter] addObserverForName:kMXSessionStateDidChangeNotification object:nil queue:[NSOperationQueue mainQueue] usingBlock:^(NSNotification *notif) {
           		MXStrongifyAndReturnIfNil(self);

                MXSession *mxSession = (MXSession*)notif.object;

                if ([[AppDelegate theDelegate].mxSessions containsObject:mxSession]
                    && mxSession.state >= MXSessionStateStoreDataReady
                    && self.shouldReceiveVoIPPushes)
                {
                    [self configurePushKit];
                    [notificationCenter removeObserver:self.matrixSessionStateObserver];                    
                }
            }];
```